### PR TITLE
JAMES-3023 Encryption for Cassandra blobStore

### PR DESCRIPTION
**Pull request by chibenwa**
Following JAMES-2921 Hybrid BlobStore we end up storing 'blobs' into Cassandra.

Given a multi-tenant environment, if I run two James servers on top of the same Cassandra database, I don't want one server to be able to read the content of the other one.

Such isolation is currently achievable by having separate keyspaces, and separate users with different right settings. However given a cassandra acls misconfiguration, such content could still be leaked from one tenant to the other one in case of James compromision.

Encrypting blobs into cassandra, with an encryption key specific per James server can be a good second line of defense to mitigate this risk. Only metatdata would be readable, raw content (headers & body) staying encrypted.

As such, I should be able to configure AES encryption of top of Cassandra storage (just like on top of ObjectStorage).

( cc @MichaelBailly )

### CODE REVIEW

**Comment by trantienduchn** [comment]
I still perfer an unified encryption configuration within the scope of BlobStore. Currently there are two: objectstorage blobstore and cassandra blobstore, it makes administrating more complex.

File: server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/EncryptedCassandraBlobStoreTest.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import org.apache.james.blob.api.BlobStore;
+import org.apache.james.blob.api.HashBlobId;
+import org.apache.james.blob.api.MetricableBlobStore;
+import org.apache.james.blob.api.MetricableBlobStoreContract;
+import org.apache.james.blob.cassandra.encryption.AesEncryptionCodec;
+import org.apache.james.blob.cassandra.encryption.EncryptionConfiguration;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class EncryptedCassandraBlobStoreTest implements MetricableBlobStoreContract {
```
Comment (trantienduchn) [inline server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/EncryptedCassandraBlobStoreTest.java]: should adding a test about these cases?
 - encrypionBlobStore read unencrypted content
 - encrypionBlobStore read another key/pass encrypted content

**Comment by chibenwa** [comment]
`I still perfer an unified encryption configuration within the scope of BlobStore.`

Given that we want to encrypt previews as well, we could be adding an (optional) cassandra component for encrypting sensible data. This would be configured in `cassandra.properties`, and can be reused straight away within 'preview encryption' scope. Would it be satisfying to you @trantienduchn ?

I am reluctant of using same credentials and keys across various storage solutions with very different mechanism and constraints...

**Comment by trantienduchn** [comment]
> This would be configured in cassandra.properties, and can be reused straight away within 'preview encryption' scope

But I'm not sure how others think about it

**Comment by mbaechler** [comment]
We tried to use tink to crypt an inputstream yesterday and we failed.
I would like we figure out a good way to do AES of our data (without loading byte arrays) before we try to spread the usage elsewhere.
BTW, if you following tink recommendation, you are supposed to generate keys randomly based on your master key and store them along the way.
It means you can't really have a stateless crypto module.
On the other end, with this design you can use different random keyset for your storage backends without having to write some specific code.

**Comment by chibenwa** [comment]
`I would like we figure out a good way to do AES of our data (without loading byte arrays)`

This is less of an issue here as we need to collect the bytes to be saved in Cassandra (no streaming is implemented via the java driver)

**Comment by chibenwa** [comment]
```
BTW, if you following tink recommendation, you are supposed to generate keys randomly based on your master key and store them along the way.
It means you can't really have a stateless crypto module.
On the other end, with this design you can use different random keyset for your storage backends without having to write some specific code.
```

This sounds like having a top level AES service, with it's own configuration, and a private key store. This sounds less than obvious to me how to be doing such a thing, and where to be storing such backend-derived-private keys, for which I see no obvious solutions (Cassandra is not an option here).

Hence having separate master keys per backend seems like a good solution to me.

**Comment by Arsnael** [comment]
```
[668672d0246ef04841ff59c132cd782e5aa90df9] [ERROR] Errors: 
[668672d0246ef04841ff59c132cd782e5aa90df9] [ERROR]   WithDefaultAwsS3Test.mailsShouldBeWellReceived » SMTPSending Error upon step D...
[668672d0246ef04841ff59c132cd782e5aa90df9] [ERROR]   WithDefaultAwsS3Test.oneHundredMailsShouldBeWellReceived » SMTPSending Error u...
[668672d0246ef04841ff59c132cd782e5aa90df9] [ERROR]   WithEncryptedAwsS3Test.mailsShouldBeWellReceived » SMTPSending Error upon step...
[668672d0246ef04841ff59c132cd782e5aa90df9] [ERROR]   WithEncryptedAwsS3Test.oneHundredMailsShouldBeWellReceived » SMTPSending Error...
```

**Comment by Arsnael** [comment]
> Duc edited:
> 
> See the fix [5772a8a](https://github.com/linagora/james-project/commit/5772a8ac1ed82cb6d9e42b962ef5175de888c7a5)

You should apply the fix here as well !

**Comment by chibenwa** [comment]
> You should apply the fix here as well !

Thanks!

**Comment by chibenwa** [comment]
`MemorySpamAssassinContractTest.deletingSpamMessageShouldNotImpactSpamAssassinState » ConditionTimeout`

test this please

**Comment by mbaechler** [comment]
> `I would like we figure out a good way to do AES of our data (without loading byte arrays)`
> 
> This is less of an issue here as we need to collect the bytes to be saved in Cassandra (no streaming is implemented via the java driver)

Streaming for Cassandra is implemented in our blobstore refactoring

**Comment by mbaechler** [comment]
> ```
> BTW, if you following tink recommendation, you are supposed to generate keys randomly based on your master key and store them along the way.
> It means you can't really have a stateless crypto module.
> On the other end, with this design you can use different random keyset for your storage backends without having to write some specific code.
> ```
> 
> This sounds like having a top level AES service, with it's own configuration, and a private key store. 

Yes it is

> This sounds less than obvious to me how to be doing such a thing, and where to be storing such backend-derived-private keys, for which I see no obvious solutions (Cassandra is not an option here).

Crypto is not meant to be easy. Why do you think storing derived keys in Cassandra is not an option?

> 
> Hence having separate master keys per backend seems like a good solution to me.

This is a conclusion I don't share. Crypto people took care to write a library to prevent us to do mistakes with crypto. And we rush to workaround this library because we think it's "good enough".

I am no crypto expert but I think we should think about it with care and only diverge from recommendations if a crypto expert tells us it's ok.

**Comment by chibenwa** [comment]
> Why do you think storing derived keys in Cassandra is not an option?

In short: storing keys on the same backend you use them to encrypt some stuff is equivalent to not encrypting this stuff...

```
read(id)
```

Is simply a bit more complicated:

```
decrypt(read(id), readPrivateKey())
```

> BTW, if you following tink recommendation, you are supposed to generate keys randomly based on your master key and store them along the way.

I would be interested if you can provide a link....

**Comment by Arsnael** [comment]
CI restart

test this please

**Comment by mbaechler** [comment]
> > Why do you think storing derived keys in Cassandra is not an option?
> 
> In short: storing keys on the same backend you use them to encrypt some stuff is equivalent to not encrypting this stuff...
> 
> ```
> read(id)
> ```
> 
> Is simply a bit more complicated:
> 
> ```
> decrypt(read(id), readPrivateKey())
> ```

Well, I think that retrieving the keyset from the database is not enough, you need both the master key and the keyset. So I think your example is too simplistic.

However, tink recommend using a dedicated system to store keys.

> > BTW, if you following tink recommendation, you are supposed to generate keys randomly based on your master key and store them along the way.
> 
> I would be interested if you can provide a link....

Look at tink doc, especially https://github.com/google/tink/blob/1.2/docs/JAVA-HOWTO.md

**Comment by chibenwa** [comment]
Not applicable

**Pull request closed without merging**